---
output:
  pdf_document:
    includes:
      in_header: header.tex
    fig_caption: false
    keep_tex: true
documentclass: article
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::knit_hooks$set(conv.odg = rmddochelper::odg.graphics.conv.hook)
```

```{r TaskCounts, echo=FALSE}
cnt <- rmddochelper::R6ClassCount$new()
cnt$set_prefix(ps_prefix = "## Problem")
```

```{r PointAssignment, echo=FALSE}
# Assign Points for Q1
lPointsQ1 <- list(TaskA = 8,
                  TaskB = 4,
                  TaskC = 2)
nPointQ1Total <- sum(unlist(lPointsQ1))
# Assign Points for Q2
lPointsQ2 <- list(TaskA = 8,
                  TaskB = 8,
                  TaskC = 3)
nPointQ2Total <- sum(unlist(lPointsQ2))
# Assign Points for Q3
lPointsQ3 <- list(TaskA = 51,
                  TaskB = 21,
                  TaskC = 0)
nPointQ3Total <- sum(unlist(lPointsQ3))
# Assign Points for Q4
lPointsQ4 <- list(TaskA = 5,
                  TaskB = 6,
                  TaskC = 2,
                  TaskD = 4)
nPointQ4Total <- sum(unlist(lPointsQ4))
# Assign Points for Q5
lPointsQ5 <- list(TaskA = 48,
                  TaskB = 21,
                  TaskC = 0)
nPointQ5Total <- sum(unlist(lPointsQ5))
# compute overal sum of points
nPointOverallTotal <- nPointQ1Total + nPointQ2Total + nPointQ3Total + nPointQ4Total + nPointQ5Total
```


\thispagestyle{empty}

\begin{tabular}{l}
ETH Zurich \\
D-USYS\\
Institute of Agricultural Sciences\\
\end{tabular}

\vspace{15ex}
\begin{center}
\huge
Post-Final Exam\\ \vspace{1ex}
Livestock Breeding and Genomics \\  \vspace{1ex}
FS 2018 \\

\normalsize
\vspace{7ex}
Peter von Rohr 
\end{center}

\vspace{7ex}
\begin{tabular}{p{5cm}lr}
  & \textsc{Date}  & \textsc{\emph{14. January 2019}} \\
  & \textsc{Begin} & \textsc{\emph{09:15 }}\\
  & \textsc{End}   & \textsc{\emph{11:15 }}\\ 
\end{tabular}

\vspace{5ex}
<!-- Table with Name -->
\large
\begin{tabular}{p{2.5cm}p{3cm}p{6cm}}
  &  Name:     &  \\
  &            &  \\
  &  Legi-Nr:  & \\
\end{tabular}
\normalsize

<!-- Table with Points -->

\vspace{9ex}
\begin{center}
\begin{tabular}{|p{3cm}|c|c|}
\hline
Problem  &  Maximum Number of Points  &  Number of Points Reached \\
\hline
1        &  `r nPointQ1Total`         & \\
\hline
2        &  `r nPointQ2Total`         & \\
\hline
3        &  `r nPointQ3Total`         & \\
\hline
4        &  `r nPointQ4Total`          & \\
\hline
5        &  `r nPointQ5Total`          & \\
\hline
Total    &  `r nPointOverallTotal`    & \\
\hline
\end{tabular}
\end{center}

\clearpage
\pagebreak




```{r, echo=FALSE, results='asis'}
cat(cnt$out(ps_suffix = "Livestock Breeding"), "\n")
```
Sheep breeders are selecting male animals for rate of gain in body weight. After one year the rams with the highest phenotypic values of the trait `rate of gain in body weight` are selected as sires of the next generation. These sires are mated to the female animals randomly. 

\begin{enumerate}
\item[a)] Do you think this is a good strategy? Please reason about your answer by providing facts from Livestock Breeding. Could you suggest some improvements to the breeding program of the sheep breeders. 
\points{`r lPointsQ1$TaskA`}
\end{enumerate}


\clearpage
\pagebreak

\begin{enumerate}
\item[b)] The sheep breeders are selecting for rate of body weight gain until all sires have a rate of gain larger than $0.8 kg$ per day. Then the breeders start to select for wool weight gain. What is the name of this selection strategy for selecting for two traits? Please specify two other selection strategies when multiple traits should be considered in a breeding program. Which strategy is the best when multiple traits should be selected for? 
\points{`r lPointsQ1$TaskB`}
\end{enumerate}


\clearpage
\pagebreak

\begin{enumerate}
\item[c)] The sheep breeders combine the traits `rate of weight gain` and `wool yield` into an aggregate genotype. The economic values of the two traits are given below. Let us assume that the two traits are combined into an index where the BLUP breeding values from an animal model are used as information sources. What are the index weights for such an selection index?
\points{`r lPointsQ1$TaskC`}
\end{enumerate}

```{r EvWoolFibre, echo=FALSE, results='asis'}
tbl_ev_wool <- tibble::data_frame(Trait = c("rate of weight gain", "wool yield"),
                                  `Economic Value` = c(0.027, 0.50))
knitr::kable(tbl_ev_wool, booktabs = TRUE, longtable = TRUE)
```


\clearpage
\pagebreak

```{r, echo=FALSE, results='asis'}
cat(cnt$out(ps_suffix = "Own Performance and Progeny Records"), "\n")
```
Given is the following data set

```{r BvPredData, echo=FALSE, results='hide'}
set.seed(1254)
n_nr_obs <- 4
### # intercept and two levels of a fixed effect
vec_beta <- c(12.55, -1.2, 2.32)
mat_x <- matrix(c(1, 1, 0,
                  1, 1, 0,
                  1, 0, 1,
                  1, 0, 1), nrow = n_nr_obs, byrow = TRUE)
### # find number of herds, knowing that first column is intercept
n_nr_herd <- ncol(mat_x)-1
### # variance components
n_sigmae2 <- 72
n_sigmaa2 <- 36
n_sigmap2 <- n_sigmaa2 + n_sigmae2
n_sigmas2 <- n_sigmaa2 / 4
n_h2 <- n_sigmaa2 / n_sigmap2
n_k_sire <- (4-n_h2)/n_h2
n_rep_t <- 0.6
n_sigmape2 <- n_rep_t * n_sigmap2 - n_sigmaa2
n_sigmate2 <- (1-n_rep_t) * n_sigmap2

### # pedigree
tbl_ped_exmp3 <- tibble::data_frame(Animal = c(4, 5, 6, 7),
                                    Sire   = c(1, 3, 3, 6),
                                    Dam    = c(2, 2, 4, 5))
### # number of animals, number of founders and number of sirees
n_nr_ani_exmp3pedext <- max(tbl_ped_exmp3$Animal)
n_nr_founder_exmp3pedext <- min(tbl_ped_exmp3$Animal) - 1
n_nr_sire_exmp3pedext <- nlevels(as.factor(tbl_ped_exmp3$Sire))
### # extend pedigree with founders
tbl_ped_exmp3_ext <- tibble::data_frame(Animal = c(1:n_nr_ani_exmp3pedext),
                                        Sire   = c(rep(NA, n_nr_founder_exmp3pedext), tbl_ped_exmp3$Sire),
                                        Dam    = c(rep(NA, n_nr_founder_exmp3pedext), tbl_ped_exmp3$Dam))

### # breeding values
### # function to get breeding values
#' @title Generate vector of breeding values 
#'
#' @description
#' Breeding values are generated using a vectorized 
#' version of the generation of breeding values
generate_vec_bv <- function(ptbl_ped, pn_sigmaa2){
  ### # get pedigree
  ped <- pedigreemm::pedigree(sire = ptbl_ped$Sire, dam = ptbl_ped$Dam, label = as.character(ptbl_ped$Animal))
  ### # number of animal
  n_nr_ani <- nrow(ptbl_ped)
  ### # get matrix D
  diag_mat_d <- diag(pedigreemm::Dmat(ped = ped), nrow = n_nr_ani, ncol = n_nr_ani)
  ### # get matrix A based on pedigree
  mat_a <- as.matrix(pedigreemm::getA(ped = ped))
  ### # cholesky of A
  mat_r <- t(chol(mat_a))
  ### # sqrt(D) to mat_s
  mat_s <- sqrt(diag_mat_d)
  ### # matrix L
  mat_l <- mat_r %*% solve(mat_s)
  ### # finally get matrix P
  mat_p <- diag(n_nr_ani) - solve(mat_l)
  ### # sample the vector of mendelian sampling
  vec_a <- rnorm(n_nr_ani, mean = 0, sd = sqrt(diag(diag_mat_d) * pn_sigmaa2))
  ### # adding pedigree Information
  vec_a_result <- mat_p %*% vec_a + vec_a
  ### # return result
  return(vec_a_result)
}
### # vector of breeding values
vec_bv <- generate_vec_bv(ptbl_ped = tbl_ped_exmp3_ext, pn_sigmaa2 = n_sigmaa2)
### # design matrix Z
mat_z <- cbind(matrix(0, nrow = n_nr_obs, ncol = n_nr_founder_exmp3pedext), diag(1, nrow = n_nr_obs, ncol = n_nr_obs))

### # generate observations
vec_y <- crossprod(t(mat_x), vec_beta) + crossprod(t(mat_z), vec_bv) + rnorm(n_nr_obs, mean = 0, sd = sqrt(n_sigmae2))

### # population mean
n_mu <- mean(vec_y)
```


```{r BvPredTable, echo=FALSE, results='asis', message=FALSE, warning=FALSE}
library(dplyr)
mat_herd <- crossprod(t(mat_x), c(0,1,2))
tbl_dat_exmp3 <- tbl_ped_exmp3 %>% mutate(Herd = mat_herd[,1], Observation = round(vec_y, digits = 2))
knitr::kable(tbl_dat_exmp3, booktabs = TRUE, longtable = TRUE)
```

The variance components $\sigma_e^2$ (residual) and $\sigma_a^2$ (genetic) are 

$$\sigma_e^2 = `r n_sigmae2`$$
$$\sigma_a^2 = `r n_sigmaa2`$$


\begin{enumerate}
\item[a)] Predict the breeding values and the corresponding accuracies of the predicted breeding values based on the animals own performance. The population mean $\mu$ is taken as the average of all observations, $\mu = `r round(n_mu, digits = 2)`$. 
\points{`r lPointsQ2$TaskA`}
\end{enumerate}



\clearpage
\pagebreak

```{r RepeatabilityModel, echo=FALSE, results='hide'}
vec_rep <- c(3, 2, 3, 3)
n_nr_obs_rep <- sum(vec_rep)

tbl_ped_exmp3_rep <- tibble::data_frame(Animal = c(rep(4, vec_rep[1]), rep(5, vec_rep[2]), rep(6, vec_rep[3]), rep(7, vec_rep[4])),
                                        Sire   = c(rep(1, vec_rep[1]), rep(3, vec_rep[2]), rep(3, vec_rep[3]), rep(6, vec_rep[4])),
                                        Dam    = c(rep(2, vec_rep[1]), rep(2, vec_rep[2]), rep(4, vec_rep[3]), rep(5, vec_rep[4])))

### # design matrix X for repeatability model
mat_x_rep <- matrix(c(rep(c(1, 1, 0), vec_rep[1]),
                      rep(c(1, 1, 0), vec_rep[2]),
                      rep(c(1, 0, 1), vec_rep[3]),
                      rep(c(1, 0, 1), vec_rep[4])), nrow = n_nr_obs_rep, byrow = TRUE)
### # design matrix W
mat_w_rep <- matrix(c(rep(c(1, 0, 0, 0), vec_rep[1]),
                      rep(c(0, 1, 0, 0), vec_rep[2]),
                      rep(c(0, 0, 1, 0), vec_rep[3]),
                      rep(c(0, 0, 0, 1), vec_rep[4])), nrow = n_nr_obs_rep, byrow = TRUE)

### # design matrix Z for repeatability model
mat_z_rep <- cbind(matrix(0, nrow = n_nr_obs_rep, ncol = n_nr_founder_exmp3pedext), mat_w_rep)

### # vector of pe
vec_pe <- rnorm(n_nr_obs, mean = 0, sd = sqrt(n_sigmape2))

### # generate observations
vec_y_rep <- crossprod(t(mat_x_rep), vec_beta) + crossprod(t(mat_z_rep), vec_bv) + crossprod(t(mat_w_rep), vec_pe) + rnorm(n_nr_obs_rep, mean = 0, sd = sqrt(n_sigmate2))

### # mean
n_mu_rep <- mean(vec_y_rep)
```

\begin{enumerate}
\item[b)] For all animals with a performance record some more observations could be collected resulting in the table of observations shown below. Predict the breeding values using the repeated observations, assuming that the repeatability $t$ is $`r n_rep_t`$. Indicate for each breeding value the corresponding accuracy.
\points{`r lPointsQ2$TaskB`}
\end{enumerate}

```{r BvPredRepTable, echo=FALSE, results='asis', message=FALSE, warning=FALSE}
mat_herd_rep <- crossprod(t(mat_x_rep), c(0,1,2))
tbl_dat_exmp3_rep <- tbl_ped_exmp3_rep %>% mutate(Herd = mat_herd_rep[,1], Observation = round(vec_y_rep, digits = 2))
knitr::kable(tbl_dat_exmp3_rep, booktabs = TRUE, longtable = TRUE)
```


\clearpage
\pagebreak


```{r ComputeMinNrObs, echo=FALSE, results='hide'}
n_min_acc <- 0.7
n_min_nr_obs_rep <- (0.7^2 * (1-n_rep_t)) / (n_h2 - 0.7^2 * n_rep_t)
```

\begin{enumerate}
\item[c)] What is the reason for the difference between the accuracies found in Problem 2a and 2b? How many observations per animal are required, if the accuracy should be greater than $`r n_min_acc`$ using the same repeatability as shown in 2b?
\points{`r lPointsQ2$TaskC`}
\end{enumerate}


\clearpage
\pagebreak

```{r, echo=FALSE, results='asis'}
cat(cnt$out(ps_suffix = "BLUP Animal Model"), "\n")
```

Given is the same data set as in Problem 2a. 

```{r, echo=FALSE, results='asis'}
knitr::kable(tbl_dat_exmp3, booktabs = TRUE, longtable = TRUE)
```

The variance components $\sigma_e^2$ (residual) and $\sigma_a^2$ (genetic) are 

$$\sigma_e^2 = `r n_sigmae2`$$
$$\sigma_a^2 = `r n_sigmaa2`$$

\begin{enumerate}
\item[a)] Predict the breeding values based on the data given above, using an animal model. Herds are treated as fixed effects and breeding values are modeled as random effects. Please specify all model components including expected values and variances for the random components. Set up the mixed model equations to get solutions for the predicted breeding values.  
\points{`r lPointsQ3$TaskA`}
\end{enumerate}


\clearpage
\pagebreak

\begin{enumerate}
\item[b)] Compute the $95\%$ confidence interval of the predicted breeding values from 3a
\points{`r lPointsQ3$TaskB`}
\end{enumerate}


\clearpage
\pagebreak

```{r, echo=FALSE, results='asis'}
cat(cnt$out(ps_suffix = "Quantitative Genetics"), "\n")
```

```{r GenerateGenoData, echo=FALSE, results='hide'}
set.seed(6789)
### # population size
n_pop_size <- 100
### # allele frequencies
n_freq_e1 <- .2
n_freq_e2 <- 1-n_freq_e1
### # genotype frequencies
n_freq_e1e1 <- n_freq_e1 ^ 2
n_freq_e1e2 <- 2 * n_freq_e1 * n_freq_e2
n_freq_e2e2 <- n_freq_e2 ^ 2
### # rounded number of genotypes
n_nr_e2e2 <- floor(n_freq_e2e2 * n_pop_size)
n_nr_e1e2 <- floor(n_freq_e1e2 * n_pop_size)
n_nr_e1e1 <- n_pop_size - n_nr_e1e2 - n_nr_e2e2
### # fix the regression coefficients
n_inter <- 7.3
n_slope <- 2.1
n_dom <- 0.8
n_res_var <- 4.2
### # vectors of values
vec_val_e1e1 <- n_inter + 2 * n_slope + rnorm(n_nr_e1e1, mean = 0, sd = sqrt(n_res_var))
vec_val_e1e2 <- n_inter + n_slope + n_dom + rnorm(n_nr_e1e2, mean = 0, sd = sqrt(n_res_var))
vec_val_e2e2 <- n_inter + rnorm(n_nr_e2e2, mean = 0, sd = sqrt(n_res_var))

tbl_geno_data_enz <- tibble::data_frame(Genotype = c(rep("$G_1G_1$", length(vec_val_e1e1)),
                                                     rep("$G_1G_2$", length(vec_val_e1e2)),
                                                     rep("$G_2G_2$", length(vec_val_e2e2))),
                                        Value = c(vec_val_e1e1, 
                                                  vec_val_e1e2, 
                                                  vec_val_e2e2))
tbl_geno_data_enz$Genotype <- as.factor(tbl_geno_data_enz$Genotype)
```

We are given a dataset where a trait of interest is mostly determined by a single genetic locus. The locus is bi-allelic with alleles $G_1$ and $G_2$. $G_1$ is the favorite allele which increases the enzyme level and it has a frequency of $`r n_freq_e1`$. For reasons of simplicity, we assume that the Hardy-Weinberg equilibrium holds. 

We take the mean levels of the trait values per genotype are taken as the characteristic values for a given genotype. This results in 

```{r MeanEnzLevel, echo=FALSE, results='asis', message=FALSE, warning=FALSE}
library(dplyr)
tbl_mean_enz <- tbl_geno_data_enz %>% group_by(Genotype) %>% summarise(`Mean Level` = round(mean(Value), digits = 2))
knitr::kable(tbl_mean_enz, booktabs = TRUE, lontable = TRUE)
```

After re-scaling the levels for each genotype such that the origin is exactly between the two homozygous genotypes $G_1G_1$ and $G_2G_2$, we get the following genotypic values for the three genotypes.

```{r RescaledEnzLevel, echo=FALSE, results='asis'}
n_scale_origin <- (tbl_mean_enz %>% filter(Genotype == "$G_1G_1$") %>% select(`Mean Level`) + tbl_mean_enz %>% filter(Genotype == "$G_2G_2$") %>% select(`Mean Level`)) / 2
n_scale_origin <- n_scale_origin[[1]]
tbl_mean_enz_resc <- tbl_mean_enz %>% mutate(`Genotypic Value` = `Mean Level` - n_scale_origin) %>% select(Genotype, `Genotypic Value`)
knitr::kable(tbl_mean_enz_resc, booktabs = TRUE, lontable = TRUE)
```



\begin{enumerate}
\item[a)] Assuming a single locus model, what are the values of $a$ and $d$ for the trait levels at locus $G$ and what is the population mean $\mu$?
\points{`r lPointsQ4$TaskA`}
\end{enumerate}


\clearpage
\pagebreak


\begin{enumerate}
\item[b)] Compute the breeding values and the dominance deviations for all three genotypes
\points{`r lPointsQ4$TaskB`}
\end{enumerate}


\clearpage
\pagebreak

\begin{enumerate}
\item[c)] Compute the genetic additive variance and the dominance variance.
\points{`r lPointsQ4$TaskC`}
\end{enumerate}


\clearpage
\pagebreak

\begin{enumerate}
\item[d)] When two heterozygous parents have offspring, these offspring individuals can have genotypes $G_1G_1$, $G_1G_2$ or $G_2G_2$. When comparing the breeding values of the offspring with different genotypes, we realize that the breeding values are different. Compute all pairwise differences between the breeding values of all offspring genotypes. What is the name of the specific effect that appears in the differences of the breeding values? 
\points{`r lPointsQ4$TaskD`}
\end{enumerate}


\clearpage
\pagebreak

```{r, echo=FALSE, results='asis'}
cat(cnt$out(ps_suffix = "Relationship and Inbreeding"), "\n")
```

Given is the following pedigree

```{r, echo=FALSE, results='asis'}
tbl_ped_exmp5 <- tibble::data_frame(Animal = c("ANTONIA",      "PHILADELPHIA", "VERA",   "PRISCA"),
                                    Sire   = c("SCHAERZ",      "PAUER",        "VITGOR", "HARRY"),
                                    Dam    = c("PHILADELPHIA", "VERA",         "PRISCA", "REIA"))

knitr::kable(tbl_ped_exmp5, booktabs = TRUE, longtable = TRUE)
```


\begin{enumerate}
\item[a)] Compute the numerator relationship matrix for the above given pedigree. 
\points{`r lPointsQ5$TaskA`}
\end{enumerate}



\clearpage
\pagebreak


\begin{enumerate}
\item[b)] Compute the inverse numerator relationship matrix $A^{-1}$ from the above given pedigree.
\points{`r lPointsQ5$TaskB`}
\end{enumerate}

