---
output:
  pdf_document:
    includes:
      in_header: header.tex
    fig_caption: false
    keep_tex: true
documentclass: article
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::knit_hooks$set(conv.odg = rmddochelper::odg.graphics.conv.hook)
```

```{r TaskCounts, echo=FALSE}
cnt <- rmddochelper::R6ClassCount$new()
cnt$set_prefix(ps_prefix = "## Problem")
```

```{r PointAssignment, echo=FALSE}
# Assign Points for Q1
lPointsQ1 <- list(TaskA = 8,
                  TaskB = 4,
                  TaskC = 2)
nPointQ1Total <- sum(unlist(lPointsQ1))
# Assign Points for Q2
lPointsQ2 <- list(TaskA = 8,
                  TaskB = 8,
                  TaskC = 3)
nPointQ2Total <- sum(unlist(lPointsQ2))
# Assign Points for Q3
lPointsQ3 <- list(TaskA = 51,
                  TaskB = 21,
                  TaskC = 0)
nPointQ3Total <- sum(unlist(lPointsQ3))
# Assign Points for Q4
lPointsQ4 <- list(TaskA = 5,
                  TaskB = 6,
                  TaskC = 2,
                  TaskD = 4)
nPointQ4Total <- sum(unlist(lPointsQ4))
# Assign Points for Q5
lPointsQ5 <- list(TaskA = 48,
                  TaskB = 21,
                  TaskC = 0)
nPointQ5Total <- sum(unlist(lPointsQ5))
# compute overal sum of points
nPointOverallTotal <- nPointQ1Total + nPointQ2Total + nPointQ3Total + nPointQ4Total + nPointQ5Total
```


\thispagestyle{empty}

\begin{tabular}{l}
ETH Zurich \\
D-USYS\\
Institute of Agricultural Sciences\\
\end{tabular}

\vspace{15ex}
\begin{center}
\huge
Solutions To Post-Final Exam\\ \vspace{1ex}
Livestock Breeding and Genomics \\  \vspace{1ex}
FS 2018 \\

\normalsize
\vspace{7ex}
Peter von Rohr 
\end{center}

\vspace{7ex}
\begin{tabular}{p{5cm}lr}
  & \textsc{Date}  & \textsc{\emph{14. January 2019}} \\
  & \textsc{Begin} & \textsc{\emph{09:15 }}\\
  & \textsc{End}   & \textsc{\emph{11:15 }}\\ 
\end{tabular}

\vspace{5ex}
<!-- Table with Name -->
\large
\begin{tabular}{p{2.5cm}p{3cm}p{6cm}}
  &  Name:     &  \\
  &            &  \\
  &  Legi-Nr:  & \\
\end{tabular}
\normalsize

<!-- Table with Points -->

\vspace{9ex}
\begin{center}
\begin{tabular}{|p{3cm}|c|c|}
\hline
Problem  &  Maximum Number of Points  &  Number of Points Reached \\
\hline
1        &  `r nPointQ1Total`         & \\
\hline
2        &  `r nPointQ2Total`         & \\
\hline
3        &  `r nPointQ3Total`         & \\
\hline
4        &  `r nPointQ4Total`          & \\
\hline
5        &  `r nPointQ5Total`          & \\
\hline
Total    &  `r nPointOverallTotal`    & \\
\hline
\end{tabular}
\end{center}

\clearpage
\pagebreak




```{r, echo=FALSE, results='asis'}
cat(cnt$out(ps_suffix = "Livestock Breeding"), "\n")
```
Sheep breeders are selecting male animals for rate of gain in body weight. After one year the rams with the highest phenotypic values of the trait `rate of gain in body weight` are selected as sires of the next generation. These sires are mated to the female animals randomly. 

\begin{enumerate}
\item[a)] Do you think this is a good strategy? Please reason about your answer by providing facts from Livestock Breeding. Could you suggest some improvements to the breeding program of the sheep breeders. 
\points{`r lPointsQ1$TaskA`}
\end{enumerate}

\sol

* The strategy is not very good, because parents do not pass phenotypes to their offspring but random samples of their alleles, hence selection should not be based on phenotypic values.
* The measurements are done only at two years of age, which is probably very late. It would be better to have observations earlier in the life of a selection candidate.
* Only male animals are measured and selected, the same should be done on the female side.
* Animals are related, this should be recorded as important information in a pedigree
* Breeding values should be predicted based on a BLUP animal model which uses all information in an optimal way
* Selection decisions should be based for males and females on the ranking according to the predicted breeding values


\clearpage
\pagebreak

\begin{enumerate}
\item[b)] The sheep breeders are selecting for rate of body weight gain until all sires have a rate of gain larger than $0.8 kg$ per day. Then the breeders start to select for wool weight gain. What is the name of this selection strategy for selecting for two traits? Please specify two other selection strategies when multiple traits should be considered in a breeding program. Which strategy is the best when multiple traits should be selected for? 
\points{`r lPointsQ1$TaskB`}
\end{enumerate}

\sol

The described strategy is called `tandem selection`

The two other strategies are 

1. Selection based on independent selection thresholds
2. Selection based on dependent selection thresholds which corresponds to a selection index using multivariate BLUP breeding values

* The best strategy is selection based on dependent selection thresholds which corresponds to selection index.


\clearpage
\pagebreak

\begin{enumerate}
\item[c)] The sheep breeders combine the traits `rate of weight gain` and `wool yield` into an aggregate genotype. The economic values of the two traits are given below. Let us assume that the two traits are combined into an index where the BLUP breeding values from an animal model are used as information sources. What are the index weights for such an selection index?
\points{`r lPointsQ1$TaskC`}
\end{enumerate}

```{r EvWoolFibre, echo=FALSE, results='asis'}
tbl_ev_wool <- tibble::data_frame(Trait = c("rate of weight gain", "wool yield"),
                                  `Economic Value` = c(0.027, 0.50))
knitr::kable(tbl_ev_wool, booktabs = TRUE, longtable = TRUE)
```

\sol

Because the same traits are used in the aggregate genotype and in the index and because the information sources used in the selection index are based on BLUP predictions, the index weights are the same as the econmic values. Hence 

$$b = w = \left[ \begin{array}{c} 0.027 \\ 0.50 \end{array} \right]$$

\clearpage
\pagebreak

```{r, echo=FALSE, results='asis'}
cat(cnt$out(ps_suffix = "Own Performance and Progeny Records"), "\n")
```
Given is the following data set

```{r BvPredData, echo=FALSE, results='hide'}
set.seed(1254)
n_nr_obs <- 4
### # intercept and two levels of a fixed effect
vec_beta <- c(12.55, -1.2, 2.32)
mat_x <- matrix(c(1, 1, 0,
                  1, 1, 0,
                  1, 0, 1,
                  1, 0, 1), nrow = n_nr_obs, byrow = TRUE)
### # find number of herds, knowing that first column is intercept
n_nr_herd <- ncol(mat_x)-1
### # variance components
n_sigmae2 <- 72
n_sigmaa2 <- 36
n_sigmap2 <- n_sigmaa2 + n_sigmae2
n_sigmas2 <- n_sigmaa2 / 4
n_h2 <- n_sigmaa2 / n_sigmap2
n_k_sire <- (4-n_h2)/n_h2
n_rep_t <- 0.6
n_sigmape2 <- n_rep_t * n_sigmap2 - n_sigmaa2
n_sigmate2 <- (1-n_rep_t) * n_sigmap2

### # pedigree
tbl_ped_exmp3 <- tibble::data_frame(Animal = c(4, 5, 6, 7),
                                    Sire   = c(1, 3, 3, 6),
                                    Dam    = c(2, 2, 4, 5))
### # number of animals, number of founders and number of sirees
n_nr_ani_exmp3pedext <- max(tbl_ped_exmp3$Animal)
n_nr_founder_exmp3pedext <- min(tbl_ped_exmp3$Animal) - 1
n_nr_sire_exmp3pedext <- nlevels(as.factor(tbl_ped_exmp3$Sire))
### # extend pedigree with founders
tbl_ped_exmp3_ext <- tibble::data_frame(Animal = c(1:n_nr_ani_exmp3pedext),
                                        Sire   = c(rep(NA, n_nr_founder_exmp3pedext), tbl_ped_exmp3$Sire),
                                        Dam    = c(rep(NA, n_nr_founder_exmp3pedext), tbl_ped_exmp3$Dam))

### # breeding values
### # function to get breeding values
#' @title Generate vector of breeding values 
#'
#' @description
#' Breeding values are generated using a vectorized 
#' version of the generation of breeding values
generate_vec_bv <- function(ptbl_ped, pn_sigmaa2){
  ### # get pedigree
  ped <- pedigreemm::pedigree(sire = ptbl_ped$Sire, dam = ptbl_ped$Dam, label = as.character(ptbl_ped$Animal))
  ### # number of animal
  n_nr_ani <- nrow(ptbl_ped)
  ### # get matrix D
  diag_mat_d <- diag(pedigreemm::Dmat(ped = ped), nrow = n_nr_ani, ncol = n_nr_ani)
  ### # get matrix A based on pedigree
  mat_a <- as.matrix(pedigreemm::getA(ped = ped))
  ### # cholesky of A
  mat_r <- t(chol(mat_a))
  ### # sqrt(D) to mat_s
  mat_s <- sqrt(diag_mat_d)
  ### # matrix L
  mat_l <- mat_r %*% solve(mat_s)
  ### # finally get matrix P
  mat_p <- diag(n_nr_ani) - solve(mat_l)
  ### # sample the vector of mendelian sampling
  vec_a <- rnorm(n_nr_ani, mean = 0, sd = sqrt(diag(diag_mat_d) * pn_sigmaa2))
  ### # adding pedigree Information
  vec_a_result <- mat_p %*% vec_a + vec_a
  ### # return result
  return(vec_a_result)
}
### # vector of breeding values
vec_bv <- generate_vec_bv(ptbl_ped = tbl_ped_exmp3_ext, pn_sigmaa2 = n_sigmaa2)
### # design matrix Z
mat_z <- cbind(matrix(0, nrow = n_nr_obs, ncol = n_nr_founder_exmp3pedext), diag(1, nrow = n_nr_obs, ncol = n_nr_obs))

### # generate observations
vec_y <- crossprod(t(mat_x), vec_beta) + crossprod(t(mat_z), vec_bv) + rnorm(n_nr_obs, mean = 0, sd = sqrt(n_sigmae2))

### # population mean
n_mu <- mean(vec_y)
```


```{r BvPredTable, echo=FALSE, results='asis', message=FALSE, warning=FALSE}
library(dplyr)
mat_herd <- crossprod(t(mat_x), c(0,1,2))
tbl_dat_exmp3 <- tbl_ped_exmp3 %>% mutate(Herd = mat_herd[,1], Observation = round(vec_y, digits = 2))
knitr::kable(tbl_dat_exmp3, booktabs = TRUE, longtable = TRUE)
```

The variance components $\sigma_e^2$ (residual) and $\sigma_a^2$ (genetic) are 

$$\sigma_e^2 = `r n_sigmae2`$$
$$\sigma_a^2 = `r n_sigmaa2`$$


\begin{enumerate}
\item[a)] Predict the breeding values and the corresponding accuracies of the predicted breeding values based on the animals own performance. The population mean $\mu$ is taken as the average of all observations, $\mu = `r round(n_mu, digits = 2)`$. 
\points{`r lPointsQ2$TaskA`}
\end{enumerate}

\sol

The predicted breeding value $\hat{a}_i$ based on an own performance is

$$\hat{a}_i = h^2(y_i - \mu)$$
where $$h^2 = \frac{\sigma_a^2}{\sigma_a^2 + \sigma_e^2} = \frac{`r n_sigmaa2`}{`r n_sigmaa2` + `r n_sigmae2`} = `r n_h2`$$

The accuracies of all predicted breeding values correspond to the correlation $r_{a, y}$ between the true breeding value and the information source $y$. 

$$r_{a,y} = \frac{cov(a,y)}{\sigma_a \sigma_y} = \frac{\sigma_a^2}{\sigma_a \sigma_y} = \frac{\sigma_a}{\sigma_y} = h = \sqrt{`r n_h2`} = `r sqrt(n_h2)`$$

The results are shown in the following table

```{r PredBvOwnPerf, echo=FALSE, results='asis'}
tbl_predbv_ownperf <- tbl_dat_exmp3 %>% 
  mutate(`Predicted Breeding Value` = n_h2 * (Observation - n_mu)) %>% 
  mutate(Accuracy = sqrt(n_h2))  %>% 
  select(Animal, `Predicted Breeding Value`, Accuracy)
knitr::kable(tbl_predbv_ownperf, booktabs = TRUE, longtable = TRUE)
```



\clearpage
\pagebreak

```{r RepeatabilityModel, echo=FALSE, results='hide'}
vec_rep <- c(3, 2, 3, 3)
n_nr_obs_rep <- sum(vec_rep)

tbl_ped_exmp3_rep <- tibble::data_frame(Animal = c(rep(4, vec_rep[1]), rep(5, vec_rep[2]), rep(6, vec_rep[3]), rep(7, vec_rep[4])),
                                        Sire   = c(rep(1, vec_rep[1]), rep(3, vec_rep[2]), rep(3, vec_rep[3]), rep(6, vec_rep[4])),
                                        Dam    = c(rep(2, vec_rep[1]), rep(2, vec_rep[2]), rep(4, vec_rep[3]), rep(5, vec_rep[4])))

### # design matrix X for repeatability model
mat_x_rep <- matrix(c(rep(c(1, 1, 0), vec_rep[1]),
                      rep(c(1, 1, 0), vec_rep[2]),
                      rep(c(1, 0, 1), vec_rep[3]),
                      rep(c(1, 0, 1), vec_rep[4])), nrow = n_nr_obs_rep, byrow = TRUE)
### # design matrix W
mat_w_rep <- matrix(c(rep(c(1, 0, 0, 0), vec_rep[1]),
                      rep(c(0, 1, 0, 0), vec_rep[2]),
                      rep(c(0, 0, 1, 0), vec_rep[3]),
                      rep(c(0, 0, 0, 1), vec_rep[4])), nrow = n_nr_obs_rep, byrow = TRUE)

### # design matrix Z for repeatability model
mat_z_rep <- cbind(matrix(0, nrow = n_nr_obs_rep, ncol = n_nr_founder_exmp3pedext), mat_w_rep)

### # vector of pe
vec_pe <- rnorm(n_nr_obs, mean = 0, sd = sqrt(n_sigmape2))

### # generate observations
vec_y_rep <- crossprod(t(mat_x_rep), vec_beta) + crossprod(t(mat_z_rep), vec_bv) + crossprod(t(mat_w_rep), vec_pe) + rnorm(n_nr_obs_rep, mean = 0, sd = sqrt(n_sigmate2))

### # mean
n_mu_rep <- mean(vec_y_rep)
```

\begin{enumerate}
\item[b)] For all animals with a performance record some more observations could be collected resulting in the table of observations shown below. Predict the breeding values using the repeated observations, assuming that the repeatability $t$ is $`r n_rep_t`$. Indicate for each breeding value the corresponding accuracy.
\points{`r lPointsQ2$TaskB`}
\end{enumerate}

```{r BvPredRepTable, echo=FALSE, results='asis', message=FALSE, warning=FALSE}
mat_herd_rep <- crossprod(t(mat_x_rep), c(0,1,2))
tbl_dat_exmp3_rep <- tbl_ped_exmp3_rep %>% mutate(Herd = mat_herd_rep[,1], Observation = round(vec_y_rep, digits = 2))
knitr::kable(tbl_dat_exmp3_rep, booktabs = TRUE, longtable = TRUE)
```

\sol

The predicted breeding value $\hat{a}_i$ based on the repeated performances is

$$\hat{a}_i = b(\tilde{y_i} - \mu)$$
where $$b = \frac{nh^2}{1+(n-1)t} \text{ and } \mu = `r round(n_mu_rep, digits = 2)`$$

The term $\tilde{y_i}$ corresponds to the mean of the repeated measures of animal $i$. 


First, we are counting the number of repetitions per animal and the mean of the repeated observations

```{r, echo=FALSE, results='asis', message=FALSE, warning=FALSE}
### # count number of repetitions per animal
tbl_dat_exmp3_nrrep_count <- tbl_dat_exmp3_rep %>%
  group_by(Animal) %>%
  summarise(`Number of Repetition` = n())

### # compute average observation of progeny per sire
tbl_dat_exmp3_meanrep <- tbl_dat_exmp3_rep %>%
  group_by(Animal) %>%
  summarise(`Mean Observation` = mean(Observation))

tbl_dat_exmp3_repmean <- tbl_dat_exmp3_meanrep %>%
  inner_join(tbl_dat_exmp3_nrrep_count)
knitr::kable(tbl_dat_exmp3_repmean, booktabs = TRUE, longtable = TRUE)
```

Then predicted breeding values are added

```{r, echo=FALSE, results='asis', message=FALSE, warning=FALSE}
tbl_dat_exmp3_bvrep <- tbl_dat_exmp3_repmean %>%
  mutate(`Predicted Breeding Value` = round(`Number of Repetition` * n_h2/(1 + (`Number of Repetition`-1) * n_rep_t) * (`Mean Observation` - n_mu_rep), digits = 3)) %>%
  mutate(Accuracy = round(sqrt(`Number of Repetition` * n_h2/(1 + (`Number of Repetition`-1) * n_rep_t)), digits = 2))
knitr::kable(tbl_dat_exmp3_bvrep)
```


\clearpage
\pagebreak


```{r ComputeMinNrObs, echo=FALSE, results='hide'}
n_min_acc <- 0.7
n_min_nr_obs_rep <- (0.7^2 * (1-n_rep_t)) / (n_h2 - 0.7^2 * n_rep_t)
```

\begin{enumerate}
\item[c)] What is the reason for the difference between the accuracies found in Problem 2a and 2b? How many observations per animal are required, if the accuracy should be greater than $`r n_min_acc`$ using the same repeatability as shown in 2b?
\points{`r lPointsQ2$TaskC`}
\end{enumerate}

\sol

* In 2a accuracies only depend on $h$ or $h^2$. 
* In 2b accuracies depend on the number of observations $n$ per animal and on the repeatability.
* The relevant inequality is

$$\sqrt{\frac{nh^2}{1+(n-1)t}} \ge 0.7$$

Squaring both sides of the inequality leads to

$$\frac{nh^2}{1+(n-1)t} \ge 0.7^2$$

Then solving for $n$ leads to

$$nh^2 \ge 0.7^2(1+(n-1)t) = 0.7^2 + 0.7^2nt - 0.7^2t$$
$$nh^2 - 0.7^2nt = n(h^2 - 0.7^2t) \ge 0.7^2 (1 - t)$$
```{r}

```

$$n \ge \frac{0.7^2 (1 - t)}{h^2 - 0.7^2t} = \frac{0.7^2 (1 - `r n_rep_t`)}{`r n_h2` - 0.7^2 * `r n_rep_t`} = `r n_min_nr_obs_rep`$$

Hence we must have at least $n = `r ceiling(n_min_nr_obs_rep)`$ observations per animal.



\clearpage
\pagebreak

```{r, echo=FALSE, results='asis'}
cat(cnt$out(ps_suffix = "BLUP Animal Model"), "\n")
```

Given is the same data set as in Problem 2a. 

```{r, echo=FALSE, results='asis'}
knitr::kable(tbl_dat_exmp3, booktabs = TRUE, longtable = TRUE)
```

The variance components $\sigma_e^2$ (residual) and $\sigma_a^2$ (genetic) are 

$$\sigma_e^2 = `r n_sigmae2`$$
$$\sigma_a^2 = `r n_sigmaa2`$$

\begin{enumerate}
\item[a)] Predict the breeding values based on the data given above, using an animal model. Herds are treated as fixed effects and breeding values are modeled as random effects. Please specify all model components including expected values and variances for the random components. Set up the mixed model equations to get solutions for the predicted breeding values.  
\points{`r lPointsQ4$TaskA`}
\end{enumerate}

\sol

### Model components
The animal model is defined as 

$$y = X \beta + Za + e$$


\begin{tabular}{lll}
where  &  $y$      &  vector of length $`r n_nr_obs`$ of observations \\
       &  $\beta$  &  vector of length $`r n_nr_herd`$ of fixed herd effects  \\
       &  $X$      &  design matrix linking fixed effects to observations \\
       &  $a$      &  vector of length $`r n_nr_ani_exmp3pedext`$ of animal breeding values \\
       &  $Z$      &  design matrix linking animal breeding values to observations \\
       &  $e$      &  vector of length $`r n_nr_obs`$ of random rest effects
\end{tabular}

The vectors of the model are

```{r animalModelComp, echo=FALSE, results='asis'}
mat_y_animal <- tbl_dat_exmp3$Observation
vec_beta_animal <- rmddochelper::vecGetVecElem(psBaseElement = "\\beta", pnVecLen = n_nr_herd, psResult = "latex")
vec_bv_animal <- rmddochelper::vecGetVecElem(psBaseElement = "a", pnVecLen = n_nr_ani_exmp3pedext, psResult = "latex")
vec_res <- rmddochelper::vecGetVecElem(psBaseElement = "e", pnVecLen = n_nr_obs, psResult = "latex")
cat("$$\n")
cat("y = \\left[")
cat(paste(rmddochelper::sConvertMatrixToLaTexArray(pmatAMatrix = mat_y_animal), collapse = "\n"), "\n")
cat("\\right] \\text{, }")
cat("\\beta = \\left[")
cat(paste(rmddochelper::sConvertVectorToLaTexArray(pvec_avector = vec_beta_animal), collapse = "\n"), "\n")
cat("\\right] \\text{, }")
cat("a = \\left[")
cat(paste(rmddochelper::sConvertVectorToLaTexArray(pvec_avector = vec_bv_animal), collapse = "\n"), "\n")
cat("\\right] \\text{, }")
cat("e = \\left[")
cat(paste(rmddochelper::sConvertVectorToLaTexArray(pvec_avector = vec_res), collapse = "\n"), "\n")
cat("\\right]")
cat("$$\n")
```

The design matrices

```{r animalModelDesignMat, echo=FALSE, results='asis'}
mat_x_model <- mat_x[,-1]
mat_z_animal <- cbind(matrix(0, nrow = n_nr_obs, ncol = n_nr_founder_exmp3pedext),
                      diag(1,nrow = n_nr_obs))
cat("$$\n")
cat("X = \\left[")
cat(paste(rmddochelper::sConvertMatrixToLaTexArray(pmatAMatrix = mat_x_model, pnDigits = 0), collapse = "\n"), "\n")
cat("\\right] \\text{, }")
cat("Z = \\left[")
cat(paste(rmddochelper::sConvertMatrixToLaTexArray(pmatAMatrix = mat_z_animal, pnDigits = 0), collapse = "\n"), "\n")
cat("\\right]")
cat("$$\n")
```



### Expected Values and Variances of Random Components

$$E\left[ \begin{array}{c} y \\ a \\ e\end{array}\right] = \left[ \begin{array}{c} X \beta \\ 0 \\ 0 \end{array}\right]$$

$$var\left[ \begin{array}{c} y \\ a \\ e\end{array}\right] = 
\left[ \begin{array}{ccc} 
          V     &  ZG  &  0 \\ 
          GZ^T  &  G   &  0 \\
          0     &  0   &  R
       \end{array}
\right]
$$
where $R = I * \sigma_e^2$, $G = A*\sigma_a^2$, $V = ZGZ^T + R$ and $A$ is the numerator relationship matrix between the animals given below  

```{r NumRelMatanimal, echo=FALSE, results='asis'}
ped_exmp3pedext <- pedigreemm::pedigree(sire = tbl_ped_exmp3_ext$Sire,
                                        dam  = tbl_ped_exmp3_ext$Dam,
                                        label = as.character(tbl_ped_exmp3_ext$Animal))
mat_a_animal <- as.matrix(pedigreemm::getA(ped = ped_exmp3pedext))

cat("$$\n")
cat("A = \\left[")
cat(paste(rmddochelper::sConvertMatrixToLaTexArray(pmatAMatrix = mat_a_animal, pnDigits = 4)))
cat("\\right]")
cat("$$\n")
```


### Mixed Model Equations
MME for the animal model are

$$
\left[
  \begin{array}{lr}
  X^TX  &  X^TZ \\
  Z^TX  &  Z^TZ + \lambda * A^{-1}
  \end{array}
\right] 
\left[
  \begin{array}{c}
  \widehat{\beta}  \\
  \widehat{a}
  \end{array}
\right]
=
\left[
  \begin{array}{c}
  X^Ty  \\
  Z^Ty
  \end{array}
\right]
$$

The single components of MME are

```{r MmeanimalComp, echo=FALSE, results='asis'}
### # coefficient matrix
n_lambda_animal <- n_sigmae2/n_sigmaa2
mat_xtx <- crossprod(mat_x_model)
mat_xtz_animal <- crossprod(mat_x_model, mat_z_animal)
mat_ztz_animal <- crossprod(mat_z_animal)
mat_ainv_animal <- solve(mat_a_animal)
mat_ztz_animal_ainv_animal_lambda <- mat_ztz_animal + n_lambda_animal * mat_ainv_animal
mat_coef_animal <- rbind(cbind(mat_xtx, mat_xtz_animal),cbind(t(mat_xtz_animal), mat_ztz_animal_ainv_animal_lambda))
### # rhs
mat_xty_animal <- crossprod(mat_x_model, mat_y_animal)
mat_zty_animal <- crossprod(mat_z_animal, mat_y_animal)
mat_rhs_animal <- rbind(mat_xty_animal, mat_zty_animal)
### # solution
mat_sol_animal <- solve(mat_coef_animal, mat_rhs_animal)

### # show the components
cat("$$\n")
cat("X^TX = \\left[ \n")
cat(paste(rmddochelper::sConvertMatrixToLaTexArray(pmatAMatrix = mat_xtx, pnDigits = 0)))
cat("\\right] \\text{, } \n")
cat("X^TZ = \\left[ \n")
cat(paste(rmddochelper::sConvertMatrixToLaTexArray(pmatAMatrix = mat_xtz_animal, pnDigits = 0)))
cat("\\right] \\text{, } \n")
cat("Z^TZ = \\left[ \n")
cat(paste(rmddochelper::sConvertMatrixToLaTexArray(pmatAMatrix = mat_ztz_animal, pnDigits = 0)))
cat("\\right] \\text{, } \n")
cat("\\lambda = \\frac{\\sigma_e^2}{\\sigma_s^2} = \\frac{", n_sigmae2, "}{", n_sigmaa2, "} = ", n_lambda_animal, sep = "")
cat("$$\n")
### # A inverse for animals
cat("$$\n")
cat("A^{-1} = \\left[ \n")
cat(paste(rmddochelper::sConvertMatrixToLaTexArray(pmatAMatrix = mat_ainv_animal, pnDigits = 4)))
cat("\\right] \n")
cat("$$\n")
```

Putting all mixed model equations together leads to

```{r MmeanimalComplete, echo=FALSE, results='asis'}
vec_beta_hat_animal <- rmddochelper::vecGetVecElem(psBaseElement = "\\widehat{\\beta}", pnVecLen = n_nr_herd, psResult = "latex")
vec_bv_hat_animal <- rmddochelper::vecGetVecElem(psBaseElement = "\\widehat{a}", pnVecLen = n_nr_ani_exmp3pedext, psResult = "latex")

cat("$$\n")
cat("\\left[ \n")
cat(paste(rmddochelper::sConvertMatrixToLaTexArray(pmatAMatrix = mat_coef_animal, pnDigits = 2)))
cat("\\right] \n")
cat("\\left[ \n")
cat(paste(rmddochelper::sConvertVectorToLaTexArray(c(vec_beta_hat_animal, vec_bv_hat_animal)), collapse = "\n"), "\n")
cat("\\right] \n")
cat("= \n")
cat("\\left[ \n")
cat(paste(rmddochelper::sConvertMatrixToLaTexArray(rbind(mat_xty_animal, mat_zty_animal), pnDigits = 2), collapse = "\n"), "\n")
cat("\\right] \n")
cat("$$\n")
```

The solution

```{r MmeanimalSol, echo=FALSE, results='asis'}
cat("$$\n")
cat("\\left[ \n")
cat(paste(rmddochelper::sConvertVectorToLaTexArray(c(vec_beta_hat_animal, vec_bv_hat_animal)), collapse = "\n"), "\n")
cat("\\right] \n")
cat("= \n")
cat("\\left[ \n")
cat(paste(rmddochelper::sConvertMatrixToLaTexArray(pmatAMatrix = mat_sol_animal, pnDigits = 4), collapse = "\n"), "\n")
cat("\\right] \n")
cat("$$\n")
```


\clearpage
\pagebreak

\begin{enumerate}
\item[b)] Compute the $95\%$ confidence interval of the predicted breeding values from 3a
\points{`r lPointsQ3$TaskB`}
\end{enumerate}

\sol

The confidence interval is based on the standard error of prediction (SEP) which is the square-root of the prediction error variance (PEV). PEV depends on the inverse of the coefficient matrix of the mixed model equations. The confidence interval is based on the assumption of the predicted breeding values being normal distributed around the current predictions with SEP corresponding to the standard deviation. The value $\alpha$ is given in the task as

$$\alpha = 1 - 0.95 = 0.05$$

```{r, echo=FALSE, results='asis'}
n_mme_dim <- nrow(mat_coef_animal)
mat_pev <- solve(mat_coef_animal)*n_sigmae2
vec_pev <- diag(mat_pev)[(n_mme_dim - n_nr_ani_exmp3pedext + 1):n_mme_dim] 
vec_sep <- sqrt(vec_pev)
alpha <- .05
vec_lower_bound <- mat_sol_animal[(n_mme_dim - n_nr_ani_exmp3pedext + 1):n_mme_dim] - qnorm(p = alpha/2, lower.tail = FALSE) * vec_sep 
vec_upper_bound <- mat_sol_animal[(n_mme_dim - n_nr_ani_exmp3pedext + 1):n_mme_dim] + qnorm(p = alpha/2, lower.tail = FALSE) * vec_sep 

### # show the table
tbl_conf_int <- tibble::data_frame(Animal = 1:n_nr_ani_exmp3pedext,
                                   PBV = mat_sol_animal[(n_mme_dim - n_nr_ani_exmp3pedext + 1):n_mme_dim],
                                   SEP = vec_sep,
                                   `Lower Bound` = vec_lower_bound,
                                   `Upper Bound` = vec_upper_bound)
knitr::kable(tbl_conf_int,
             booktabs = TRUE,
             longtable = TRUE)
```




\clearpage
\pagebreak

```{r, echo=FALSE, results='asis'}
cat(cnt$out(ps_suffix = "Quantitative Genetics"), "\n")
```

```{r GenerateGenoData, echo=FALSE, results='hide'}
set.seed(6789)
### # population size
n_pop_size <- 100
### # allele frequencies
n_freq_e1 <- .2
n_freq_e2 <- 1-n_freq_e1
### # genotype frequencies
n_freq_e1e1 <- n_freq_e1 ^ 2
n_freq_e1e2 <- 2 * n_freq_e1 * n_freq_e2
n_freq_e2e2 <- n_freq_e2 ^ 2
### # rounded number of genotypes
n_nr_e2e2 <- floor(n_freq_e2e2 * n_pop_size)
n_nr_e1e2 <- floor(n_freq_e1e2 * n_pop_size)
n_nr_e1e1 <- n_pop_size - n_nr_e1e2 - n_nr_e2e2
### # fix the regression coefficients
n_inter <- 7.3
n_slope <- 2.1
n_dom <- 0.8
n_res_var <- 4.2
### # vectors of values
vec_val_e1e1 <- n_inter + 2 * n_slope + rnorm(n_nr_e1e1, mean = 0, sd = sqrt(n_res_var))
vec_val_e1e2 <- n_inter + n_slope + n_dom + rnorm(n_nr_e1e2, mean = 0, sd = sqrt(n_res_var))
vec_val_e2e2 <- n_inter + rnorm(n_nr_e2e2, mean = 0, sd = sqrt(n_res_var))

tbl_geno_data_enz <- tibble::data_frame(Genotype = c(rep("$G_1G_1$", length(vec_val_e1e1)),
                                                     rep("$G_1G_2$", length(vec_val_e1e2)),
                                                     rep("$G_2G_2$", length(vec_val_e2e2))),
                                        Value = c(vec_val_e1e1, 
                                                  vec_val_e1e2, 
                                                  vec_val_e2e2))
tbl_geno_data_enz$Genotype <- as.factor(tbl_geno_data_enz$Genotype)
```

We are given a dataset where a trait of interest is mostly determined by a single genetic locus. The locus is bi-allelic with alleles $G_1$ and $G_2$. $G_1$ is the favorite allele which increases the enzyme level and it has a frequency of $`r n_freq_e1`$. For reasons of simplicity, we assume that the Hardy-Weinberg equilibrium holds. 

We take the mean levels of the trait values per genotype are taken as the characteristic values for a given genotype. This results in 

```{r MeanEnzLevel, echo=FALSE, results='asis', message=FALSE, warning=FALSE}
library(dplyr)
tbl_mean_enz <- tbl_geno_data_enz %>% group_by(Genotype) %>% summarise(`Mean Level` = round(mean(Value), digits = 2))
knitr::kable(tbl_mean_enz, booktabs = TRUE, lontable = TRUE)
```

After re-scaling the levels for each genotype such that the origin is exactly between the two homozygous genotypes $G_1G_1$ and $G_2G_2$, we get the following genotypic values for the three genotypes.

```{r RescaledEnzLevel, echo=FALSE, results='asis'}
n_scale_origin <- (tbl_mean_enz %>% filter(Genotype == "$G_1G_1$") %>% select(`Mean Level`) + tbl_mean_enz %>% filter(Genotype == "$G_2G_2$") %>% select(`Mean Level`)) / 2
n_scale_origin <- n_scale_origin[[1]]
tbl_mean_enz_resc <- tbl_mean_enz %>% mutate(`Genotypic Value` = `Mean Level` - n_scale_origin) %>% select(Genotype, `Genotypic Value`)
knitr::kable(tbl_mean_enz_resc, booktabs = TRUE, lontable = TRUE)
```



\begin{enumerate}
\item[a)] Assuming a single locus model, what are the values of $a$ and $d$ for the trait levels at locus $G$ and what is the population mean $\mu$?
\points{`r lPointsQ4$TaskA`}
\end{enumerate}

\pagebreak

\sol

The genotypic values given in the last table of the problem corresponds to the following symbolic values

```{r GenotypicValue, echo=FALSE, results='asis'}
tbl_geno_val <- tbl_mean_enz_resc %>% 
  select(Genotype) %>% 
  mutate(`Genotypic Value` = c("$V_{11} = a$", "$V_{12} = d$", "$V_{22} = -a$"))
knitr::kable(tbl_geno_val, booktabs = TRUE, longtable = TRUE)
```

Hence 

```{r ValueAD, echo=FALSE, results='hide'}
n_value_a <- tbl_mean_enz_resc %>% filter(Genotype == "$G_1G_1$") %>% select(`Genotypic Value`)
n_value_d <- tbl_mean_enz_resc %>% filter(Genotype == "$G_1G_2$") %>% select(`Genotypic Value`)
```

$$a = `r n_value_a`$$
and
$$d = `r n_value_d`$$

The population mean is defined as 

$$\mu = V_{11} * f(G_1G_1) + V_{12} * f(G_1G_2) + V_{22} * f(G_2G_2)$$

where $f(G_iG_j)$ stands for the frequency of genotype $G_iG_j$. From Hardy-Weinberg we know that 

```{r GenoFreq, echo=FALSE, results='asis'}
tbl_geno_freq <- tbl_mean_enz_resc %>% 
  select(Genotype) %>% 
  mutate(`Genotype Frequency` = c("$f(G_1G_1) = p^2$", "$f(G_1G_2) = 2pq$", "$f(G_2G_2) = q^2$"))
knitr::kable(tbl_geno_freq, booktabs = TRUE, longtable = TRUE)
```

where $p = f(G_1)$ the allele frequency of $G_1$ which is given in the task and is $p = `r n_freq_e1`$. Because the locus $G$ is bi-allelic, we get $q = f(G_2) = 1-p = `r n_freq_e2`$.

Using all this, we get the population mean $\mu$ as

```{r PopMean, echo=FALSE, results='hide'}
n_pop_mean <- (n_freq_e1 - n_freq_e2) * n_value_a + 2 * n_freq_e1 * n_freq_e2 * n_value_d
```


$$\mu = a * p^2 + d * 2pq + (-a) * q^2 = (p-q)a + 2pqd = (`r n_freq_e1` - `r n_freq_e2`) * `r n_value_a` + 2 * `r n_freq_e1` * `r n_freq_e2` * `r n_value_d` = `r n_pop_mean`$$

\clearpage
\pagebreak


\begin{enumerate}
\item[b)] Compute the breeding values and the dominance deviations for all three genotypes
\points{`r lPointsQ4$TaskB`}
\end{enumerate}

\sol

```{r BreedingValues, echo=FALSE, results='hide'}
n_value_alpha <- n_value_a + (n_freq_e2 - n_freq_e1) * n_value_d
n_bv_e1e1 <- 2 * n_freq_e2 * n_value_alpha
n_bv_e1e2 <- (n_freq_e2 - n_freq_e1) * n_value_alpha
n_bv_e2e2 <- -2 * n_freq_e1 * n_value_alpha
```

All breeding values depend on the allele substitution effect $\alpha$ which corresponds to 

$$\alpha = a + (q-p)d = `r n_value_a` + (`r n_freq_e2` - `r n_freq_e1`) * `r n_value_d` = `r n_value_alpha`$$

Breeding values are computed according to the following table

```{r BvTable, echo=FALSE, results='asis'}
tbl_bv <- tbl_mean_enz_resc %>% 
  select(Genotype) %>% 
  mutate(`Breeding Value` = c(paste0("$2q \\alpha = ", n_bv_e1e1, "$"),
                              paste0("$(q-p) \\alpha = ", n_bv_e1e2, "$"),
                              paste0("$-2p \\alpha = ", n_bv_e2e2, "$")))
knitr::kable(tbl_bv, booktabs = TRUE, longtable = TRUE)
```

The dominance deviation is

```{r DomDev, echo=FALSE, results='asis'}
n_dom_e1e1 <- -2 * n_freq_e2^2 * n_value_d
n_dom_e1e2 <- 2 * n_freq_e1 * n_freq_e2 * n_value_d
n_dom_e2e2 <- -2 * n_freq_e1^2 * n_value_d
tbl_dom <- tbl_mean_enz_resc %>% 
  select(Genotype) %>% 
  mutate(`Dominance Deviation` = c(paste0("$-2q^2d = ", n_dom_e1e1, "$"),
                                   paste0("$2pqd = ", n_dom_e1e2, "$"),
                                   paste0("$-2p^2d = ", n_dom_e2e2, "$")))
knitr::kable(tbl_dom, booktabs = TRUE, longtable = TRUE)
```


\clearpage
\pagebreak

\begin{enumerate}
\item[c)] Compute the genetic additive variance and the dominance variance.
\points{`r lPointsQ4$TaskC`}
\end{enumerate}

\sol

The additive genetic variance corresponds to 

```{r GeneticVariances, echo=FALSE, results='hide'}
n_sigma_a2 <- 2 * n_freq_e1 * n_freq_e2 * n_value_alpha^2
n_sigma_d2 <- (2 * n_freq_e1 * n_freq_e2 * n_value_d)^2
```

$$\sigma_A^2 = 2pq\alpha^2 = 2 * `r n_freq_e1` * `r n_freq_e2` * `r n_value_alpha`^2 = `r n_sigma_a2`$$

$$\sigma_D^2 = (2pqd)^2 = (2 * `r n_freq_e1` * `r n_freq_e2` * `r n_value_d`)^2 = `r n_sigma_d2`$$


\clearpage
\pagebreak

\begin{enumerate}
\item[d)] When two heterozygous parents have offspring, these offspring individuals can have genotypes $G_1G_1$, $G_1G_2$ or $G_2G_2$. When comparing the breeding values of the offspring with different genotypes, we realize that the breeding values are different. Compute all pairwise differences between the breeding values of all offspring genotypes. What is the name of the specific effect that appears in the differences of the breeding values? 
\points{`r lPointsQ4$TaskD`}
\end{enumerate}

\sol

$$BV_{11} - BV_{12} = 2q\alpha - (q-p)\alpha = \alpha$$
$$BV_{12} - BV_{22} = (q-p)\alpha - (-2p\alpha) = \alpha$$
$$BV_{11} - BV_{22} = 2q\alpha - (-2p\alpha) = 2\alpha$$

The effect $\alpha$ appears in all differences and is called __allele substitution__ effect.

\clearpage
\pagebreak

```{r, echo=FALSE, results='asis'}
cat(cnt$out(ps_suffix = "Relationship and Inbreeding"), "\n")
```

Given is the following pedigree

```{r, echo=FALSE, results='asis'}
tbl_ped_exmp5 <- tibble::data_frame(Animal = c("ANTONIA",      "PHILADELPHIA", "VERA",   "PRISCA"),
                                    Sire   = c("SCHAERZ",      "PAUER",        "VITGOR", "HARRY"),
                                    Dam    = c("PHILADELPHIA", "VERA",         "PRISCA", "REIA"))

knitr::kable(tbl_ped_exmp5, booktabs = TRUE, longtable = TRUE)
```


\begin{enumerate}
\item[a)] Compute the numerator relationship matrix for the above given pedigree. 
\points{`r lPointsQ5$TaskA`}
\end{enumerate}

\sol

First we have to re-order the pedigree such that parents are always before progeny. This results in


```{r, echo=FALSE, results='asis'}
tbl_ped_exmp5_reord <- tibble::data_frame(Animal = rev(tbl_ped_exmp5$Animal),
                                    Sire   = rev(tbl_ped_exmp5$Sire),
                                    Dam    = rev(tbl_ped_exmp5$Dam))

knitr::kable(tbl_ped_exmp5_reord, booktabs = TRUE, longtable = TRUE)
```


Recoding the pedigree to numers with the following mapping

```{r, echo=FALSE, results='asis'}
tbl_ped_exmp5_map <- tibble::data_frame(Name = c("HARRY","REIA","VITGOR","PAUER","SCHAERZ","PRISCA","VERA", "PHILADELPHIA", "ANTONIA"),
                                        Code = c(1:9))
knitr::kable(tbl_ped_exmp5_map, booktabs = TRUE, longtable = TRUE)
```

Applying the above shown mapping

```{r, echo=FALSE, results='asis'}
tbl_ped_exmp5_recode <- tibble::data_frame(Animal = c(6, 7, 8, 9),
                                    Sire   = c(1, 3, 4, 5),
                                    Dam    = c(2, 6, 7, 8))
knitr::kable(tbl_ped_exmp5_recode, booktabs = TRUE, longtable = TRUE)
```

Extending the pedigree with founders

```{r, echo=FALSE, results='asis'}
n_nr_founder_emp5 <- min(tbl_ped_exmp5_recode$Animal)-1
n_nr_ani_emp5 <- max(tbl_ped_exmp5_recode$Animal)
tbl_ped_exmp5_ext <- tibble::data_frame(Animal = c(1:n_nr_ani_emp5),
                                        Sire   = c(rep(NA, n_nr_founder_emp5), tbl_ped_exmp5_recode$Sire),
                                        Dam    = c(rep(NA, n_nr_founder_emp5), tbl_ped_exmp5_recode$Dam))
knitr::kable(tbl_ped_exmp5_ext, booktabs = TRUE, longtable = TRUE)
```

This pedigree can be used to compute the numerator relationship matrix

```{r, echo=FALSE, results='asis'}
ped_exmp5 <- pedigreemm::pedigree(sire = tbl_ped_exmp5_ext$Sire,
                                  dam  = tbl_ped_exmp5_ext$Dam,
                                  label = as.character(tbl_ped_exmp5_ext$Animal))
mat_a_exmp5 <- as.matrix(pedigreemm::getA(ped = ped_exmp5))

cat("$$\n")
cat("A = \\left[ \n")
cat(paste(rmddochelper::sConvertMatrixToLaTexArray(pmatAMatrix = mat_a_exmp5, pnDigits = 4), collapse = "\n"), "\n")
cat("\\right] \n")
cat("$$\n")
```


\clearpage
\pagebreak


\begin{enumerate}
\item[b)] Compute the inverse numerator relationship matrix $A^{-1}$ from the above given pedigree.
\points{`r lPointsQ5$TaskB`}
\end{enumerate}

\sol

```{r, echo=FALSE, results='asis'}
mat_ainv_exmp5 <- as.matrix(pedigreemm::getAInv(ped = ped_exmp5))

cat("$$\n")
cat("A^{-1} = \\left[ \n")
cat(paste(rmddochelper::sConvertMatrixToLaTexArray(pmatAMatrix = mat_ainv_exmp5, pnDigits = 4), collapse = "\n"), "\n")
cat("\\right] \n")
cat("$$\n")
```


